### ISSUE TITLE
Player checking in Android via `pm list packages`

### ISSUE THREAD

**Issue by Rem01Gaming** [pystardust/ani-cli#1416]
**Is your feature request related to a problem? Please describe.**
There is a way to check players in Android, we can retrieve package list from command `pm list packages` and then check if MPV or VLC is installed.

**Describe the solution you'd like**
Example of checking installed player in Android:
```shell
if pm list packages | grep is.xyz.mpv >/dev/null; then
	echo "MPV Android installed"
fi

if pm list packages | grep org.videolan.vlc >/dev/null; then
	echo "VLC Android installed"
fi
```

**Additional context**
Instead of completely disabling the player checking on Android, this is much more better solution.

**Comment by CoolnsX** [comment]
You think I haven't tried that? I know that it won't work.
![Screenshot_20240908-190403](https://github.com/user-attachments/assets/6dc4bbfa-3dbc-46b0-95bd-011b126906d4)

**Comment by Rem01Gaming** [comment]
Hmm try `/system/bin/pm list packages`

**Comment by CoolnsX** [comment]
Interesting, it works now, earlier it used to not work. The same binary path u mentioned.

**Comment by CoolnsX** [comment]
I have Android 14 btw

**Comment by Rem01Gaming** [comment]
> I have Android 14 btw

Android 14 is a bit weird on some case like this
In case it still failing, we can disables the player checking only if the error pm command returns an error.

**Comment by CoolnsX** [comment]
Nah the first pm is from termux. And this one is from system.

**Comment by Rem01Gaming** [comment]
> Ye, the first pm is from termux. And this one is from system.

![Screenshot_20240908-204926_Termux_Monet~2](https://github.com/user-attachments/assets/718a963f-616b-4853-a2e7-cf6383f13aa0)
It's actually the same pm from system, just how to call it that diff

**Comment by CoolnsX** [comment]
That calling via termux's pm is not working properly.

**Comment by Rem01Gaming** [comment]
Ye

**Comment by port19x** [comment]
@CoolnsX wanna do sth with this or should we omit this?

**Comment by CoolnsX** [comment]
We need more devices to confirm this, we can implement it but it should not be blocking

**Comment by Nannk** [comment]
LineageOS 21 with microG, Android 14 (AP2A.240905.003)
![Screenshot_20241017-110808_Termux_1](https://github.com/user-attachments/assets/5b7a0245-a84e-433b-a09a-361269449285)

**Comment by Rem01Gaming** [comment]
> You think I haven't tried that? I know that it won't work. ![Screenshot_20240908-190403](https://private-user-images.githubusercontent.com/76195824/365441682-6dc4bbfa-3dbc-46b0-95bd-011b126906d4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkxNzc3ODAsIm5iZiI6MTcyOTE3NzQ4MCwicGF0aCI6Ii83NjE5NTgyNC8zNjU0NDE2ODItNmRjNGJiZmEtM2RiYy00NmIwLTk1YmQtMDExYjEyNjkwNmQ0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE3VDE1MDQ0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFlYzkwMWVmZGM4OWQ2ZWQ3YTkxY2I3OTY5OTkwOWI4OWRjNzE0MGI2NGI1NjhjZjE1NGE2ODA0OGFlYmRkZWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.68krhbId_nes8DTMbLo_Hc5o0Qxx3kJmZKkUx0Nng-g)

@CoolnsX I think you're right, after some research Google have been restrict `QUERY_ALL_PACKAGES` permission access [since Android 11](https://medium.com/androiddevelopers/package-visibility-in-android-11-cc857f221cd9).
I'm also found this [discussion on termux packages repo](https://github.com/termux/termux-packages/discussions/8292) which talking about the same thing, however somehow `pm list packages` working some ROMs (yes it's not device dependent). About why it was working on some ROMs is either device maintainer or ROM developer have been bypassed or skipped this particular security changes.

Since this was working on some ROMs, I recommend to disables the player checking only if the `pm` command returns an error.
```shell
androidPlayerCheck() {
	package_list="$(/system/bin/pm list packages)"

	if echo "$package_list" | grep -q Failure; then
		echo "Checking of players is disabled due to selinux restriction"
		return 1
	fi

	if echo "$package_list" | grep -q is.xyz.mpv; then
		echo "MPV Android installed"
		return 0
	fi

	if echo "$package_list" | grep -q org.videolan.vlc; then
		echo "VLC Android installed"
		return 0
	fi
}
androidPlayerCheck
```

But yeah we probably need to think twice to implement this, it's up to you.

**Comment by searinminecraft** [comment]
I think the reason why `/system/bin/pm list packages` doesnt work is because of SELinux related stuff. It depends on kernel or SELinux status or rules

**Comment by SamRezaXYZ** [comment]
Not working on android 14+ ?

**Comment by Rem01Gaming** [comment]
> Not working on android 14+ ?

Idk, sometimes it works sometimes not
Depending the ROM

**Comment by SamRezaXYZ** [comment]
> > Not working on android 14+ ?
> 
> Idk, sometimes it works sometimes not Depending the ROM

What roms you tried ?

**Comment by SamRezaXYZ** [comment]
> > Not working on android 14+ ?
> 
> Idk, sometimes it works sometimes not Depending the ROM

![Screenshot_2025-01-11-10-28-58-679_com termux-edit](https://github.com/user-attachments/assets/c444ac26-dc8a-4d8d-bd37-df2cd3439e63)

I just don't know how to enable vlc or mpv


Using HyperOS 1

**Comment by Rem01Gaming** [comment]
> I think the reason why `/system/bin/pm list packages` doesnt work is because of SELinux related stuff. It depends on kernel or SELinux status and rules
> 
> It really only works if there is a workaround to get around the SELinux restriction, or by setting the SELinux to Permissive (which is a security flaw that is not recommended to do, and requires root)

Maybe we could run `pm list` if ROOT available...

We check the ROOT first if it's available, if ROOT available we can run `su -c pm list` and return package list as intended.
If ROOT is not available just run `/system/bin/pm list` anyway and see if it's returns any errors, when error happening disable player checking.

**Comment by Rem01Gaming** [comment]
> What roms you tried ?
I'm using DotOS A12L rn, some users from variety of ROMs response differently.

**Comment by Derisis13** [comment]
> Maybe we could run `pm list` if ROOT available...
> 
> We check the ROOT first if it's available, if ROOT available we can run `su -c pm list` and return package list as intended. If ROOT is not available just run `/system/bin/pm list` anyway and see if it's returns any errors, when error happening disable player checking.

At that point, telling the user what player they are missing is pretty much irrelevant, if someone can root their phone, they should be ablr to install ani-cli properly.

**Comment by Rem01Gaming** [comment]
> At that point, telling the user what player they are missing is pretty much irrelevant, if someone can root their phone, they should be ablr to install ani-cli properly.

After some thoughts I kind of agree with this, maybe we should start caring about other important things...

**Comment by port19x** [comment]
I agree with @Derisis13 on this one. If this doesn't work reliably on all androids we can just skip this

**Comment by Rem01Gaming** [comment]
Okay
